File: src/matt/media/player/MediaHandle.java
package matt.media.player;

public abstract class MediaHandle
{
	protected MediaHandle prev, next;
	
	public MediaHandle(MediaHandle prev, MediaHandle next)
	{
		this.prev = prev;
		if(prev != null)
			prev.next = this;
		this.next = next;
		if(next != null)
			next.prev = this;
	}
	
	public MediaHandle(MediaHandle prev)
	{
		this(prev, null);
	}
	
	public abstract AudioSource getCurrentAudioSource();
	public abstract AudioSource getAudioSource(int index);
	public abstract Playlist getPlaylist();
	
	public MediaHandle getNext()
	{
		return next;
	}
	
	public MediaHandle getPrev()
	{
		return prev;
	}
	
	public MediaHandle setNext(MediaHandle newMediaHandle)
	{
		MediaHandle old = next;
		next = newMediaHandle;
		if(next != null)
			next.prev = this;
		if(old != null)
			old.prev = null;
		return old;
	}
	
	public MediaHandle setPrev(MediaHandle newMediaHandle)
	{
		MediaHandle old = prev;
		prev = newMediaHandle;
		if(prev != null)
			prev.next = this;
		if(old != null)
			old.next = null;
		return old;
	}
}

File: src/matt/media/player/MediaHandle.kt
package matt.media.player

sealed class MediaHandle
{
    abstract fun getCurrentAudioSource(): AudioSource
    abstract fun getAudioSource(index: Int): AudioSource
    abstract fun getPlaylist(): Playlist
}

class SongHandle(private val audioSource: AudioSource): MediaHandle()
{
    override fun getCurrentAudioSource() = audioSource
    override fun getAudioSource(index: Int) = if(index == 0) audioSource else throw IllegalArgumentException("AudioSource from SongHandle cannot be accessed with non-zero index")
    override fun getPlaylist() = throw UnsupportedOperationException("Cannot get playlist from SongHandle")
}

class PlaylistHandle(private val playlistSource: Playlist): MediaHandle()
{
    override fun getCurrentAudioSource() = throw UnsupportedOperationException("Cannot get current AudioSource from PlaylistHandle")
    override fun getAudioSource(index: Int) = playlistSource.getSong(index).getCurrentAudioSource()
    override fun getPlaylist() = playlistSource
}